feat: add `--cache-from` container build flag in publish

RanadeepPolavarapu opened the pull request:
### Checklist

<!--
	thank you for contributing to CircleCI Orbs!
	before submitting your a request, please go through the following
	items and place an x in the [ ] if they have been completed
-->

- [X] All new jobs, commands, executors, parameters have descriptions
- [X] Examples have been added for any significant new features
- [X] README has been updated, if necessary

### Motivation, issues

<!---
	why is this change required? what problem does it solve?
  paste links to any relevant GitHub issues filed against
  this repository that this pull request addresses
-->
Allows the publish job to get an optional performance boost by using Docker's `--cache-from` flag.

### Description

<!---
  Describe your changes in detail, preferably in an imperative mood,
  i.e., "add `commandA` to `jobB`"
 -->
Added `--cache-from` to container build in publish job.

Resolves #15 .

RanadeepPolavarapu commented:
Added an implicit `docker pull` to fetch the cache for `--cache-from` flag. Please let me know if you prefer not to have this. I thought this would be a nice-to-have to avoid users from having to do an explicit `docker pull` themselves when they add the `cache-from` param.

KyleTryon commented on src/commands/build.yml:
```diff
@@ -29,8 +29,8 @@ parameters:
-     If linting Dockerfile, treat linting warnings as errors (would trigger
-     an exist code and fail the CircleCI job)?
+      If linting Dockerfile, treat linting warnings as errors (would trigger
+      an exist code and fail the CircleCI job)?
```
This should be `exit` code.

KyleTryon commented on src/jobs/publish.yml:
```diff
@@ -39,8 +39,8 @@ parameters:
-     If linting Dockerfile, treat linting warnings as errors (would trigger
-     an exist code and fail the CircleCI job)?
+      If linting Dockerfile, treat linting warnings as errors (would trigger
+      an exist code and fail the CircleCI job)?
```
`exit code`

KyleTryon commented on src/commands/build.yml:
```diff
@@ -72,7 +81,13 @@ steps:
+        done
+
         docker build \
+          --cache-from "<<parameters.cache-from>>" \
```
While I appreciate the intent I am not sure this makes much sense for CircleCI. On the container you are currently building on, no other images will be present in the container to use as cache unless you had already previously pulled them in the same job.

RanadeepPolavarapu replied:
I posted a comment related to this in CircleCI-Public/aws-ecr-orb#4. I am open to input, shall we continue the discussion there to avoid double posting?

RanadeepPolavarapu commented:
@KyleTryon I changed the parameter to use underscore case `cache_from` instead of `cache-from`. Also, added into the param description about the implicit `docker pull`.

Also, attributed @vsoch as a co-author. If it is easier to merge this then, please feel free. 

The git gimmicks were too difficult to maintain by making PR to someone-else's PR. 😄

vsoch commented:
Yeah sorry @RanadeepPolavarapu - I wanted it to work but I’m so hugely uncomfortable with anything that isn’t clean commit push, etc. (cherry picks and similar). For here what I’ll do for a final clean up of our mess is a git rebase with master, if possible.

RanadeepPolavarapu commented:
@vsoch Don't sweat it, I totally understand 😄 . Yes, agreed, it was a bit hard to maintain with the PR on top of PR. My fault, apologies for it.

Beanow commented on src/commands/build.yml:
```diff
@@ -47,6 +47,17 @@ parameters:
+    description: >
+      Images to consider as cache sources. Specified as comma separated values of
+      image and tag syntax. An implicit `docker pull` will be done for each image
+      specific to fetch thei mage. Thus, an explicit `docker pull` does not need
```
`thei` > `their`

Beanow commented:
Overall think this should do the job :+1: well job

RanadeepPolavarapu commented:
Fixed typo suggestions from @Beanow and squashed all commits into one.

Beanow commented on src/commands/build.yml:
```diff
@@ -72,7 +83,13 @@ steps:
+        done
+
         docker build \
+          --cache-from "<<parameters.cache_from>>" \
```
Looking at #27 doesn't this require a condition and unless version as well?

RanadeepPolavarapu replied:
The quotes around it take care of that. Docker allows an empty —cache-from which means just go to disk and use all the cache layers available. Since the default value of the param is empty string, it’s equal to `—cache-from “”`.

Beanow commented:
Closer look now I saw #27

lokst commented:
@RanadeepPolavarapu After consideration, I went with merging #27 and hence will close this PR due to the functionality overlap between the two PRs. Thank you for your work on this feature and apologies for the delay in reviewing!
